// syncano-server/lib/server.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _data = require('./data');

var _data2 = _interopRequireDefault(_data);

var _users = require('./users');

var _users2 = _interopRequireDefault(_users);

var _account = require('./account');

var _account2 = _interopRequireDefault(_account);

var _instance = require('./instance');

var _instance2 = _interopRequireDefault(_instance);

var _event = require('./event');

var _event2 = _interopRequireDefault(_event);

var _socket = require('./socket');

var _socket2 = _interopRequireDefault(_socket);

var _response = require('./response');

var _response2 = _interopRequireDefault(_response);

var _logger = require('./logger');

var _logger2 = _interopRequireDefault(_logger);

var _channel = require('./channel');

var _channel2 = _interopRequireDefault(_channel);

var _class2 = require('./class');

var _class3 = _interopRequireDefault(_class2);

var _settings = require('./settings');

var _settings2 = _interopRequireDefault(_settings);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var server = function server() {
  var ctx = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  var settings = new _settings2.default(ctx);
  var getConfig = function getConfig(className) {
    return Object.assign({ className: className }, settings);
  };
  var config = getConfig();

  var _class = new _class3.default(config);
  var users = new _users2.default(config);
  var event = new _event2.default(config);
  var channel = new _channel2.default(config);
  var socket = new _socket2.default(config);
  var response = new _response2.default(config);
  var account = new _account2.default(config);
  var instance = new _instance2.default(config);
  var logger = new _logger2.default(config);

  return {
    _class: _class,
    users: users,
    account: account,
    instance: instance,
    event: event,
    channel: channel,
    socket: socket,
    response: response,
    logger: logger,
    data: new Proxy(new _data2.default(settings), {
      get: function get(target, className) {
        return new _data2.default(getConfig(className));
      }
    })
  };
};

exports.default = server;
module.exports = exports['default'];
